# Puzzle 3

--- Día 3: Diagnóstico binario ---
El submarino ha estado haciendo algunos crujidos extraños, así que le pides que produzca un informe de diagnóstico por si acaso.

El informe de diagnóstico (su entrada del rompecabezas) consiste en una lista de números binarios que, cuando se decodifican correctamente, pueden decirle muchas cosas útiles sobre las condiciones del submarino. El primer parámetro a comprobar es el consumo de energía.

Debe usar los números binarios en el informe de diagnóstico para generar dos nuevos números binarios (llamados tasa gamma y tasa épsilon). El consumo de energía se puede calcular multiplicando la tasa gamma por la tasa épsilon.

Cada bit de la velocidad gamma se puede determinar encontrando el bit más común en la posición correspondiente de todos los números en el informe de diagnóstico. Por ejemplo, dado el siguiente informe de diagnóstico:

00100

11110

10110

10111

10101

01111

00111

11100

10000

11001

00010

01010

Considerando solo el primer bit de cada número, hay cinco bits 0 y siete bits 1. Dado que el bit más común es 1, el primer bit de la velocidad gamma es 1.

El segundo bit más común de los números en el informe de diagnóstico es 0, por lo que el segundo bit de la tasa de gamma es 0.

El valor más común del tercer, cuarto y quinto bits son 1, 1 y 0, respectivamente, por lo que los tres últimos bits de la velocidad gamma son 110.

Entonces, la tasa gamma es el número binario 10110, o 22 en decimal.

La tasa de épsilon se calcula de manera similar; en lugar de usar el bit más común, se usa el bit menos común de cada posición. Entonces, la tasa de épsilon es 01001, o 9 en decimal. Multiplicar la tasa gamma (22) por la tasa épsilon (9) produce el consumo de energía, 198.

Utilice los números binarios en su informe de diagnóstico para calcular la tasa gamma y la tasa épsilon, luego multiplíquelos. ¿Cuál es el consumo de energía del submarino? (Asegúrese de representar su respuesta en decimal, no en binario).

In [25]:
def binario_a_decimal(binario):
    posicion = 0
    decimal = 0
    # Invertir la cadena porque debemos recorrerla de derecha a izquierda
    binario = binario[::-1]
    for digito in binario:
        # Elevar 2 a la posición actual
        multiplicador = 2**posicion
        decimal += int(digito) * multiplicador
        posicion += 1
    return decimal


In [32]:
contenido = open('input_puzzle3.txt', 'r')
mensaje = contenido.readlines()
contador_cero = 0
contador_uno = 0
gamma = ''
epsilon = ''
for i in range(12):
    for linea in mensaje:
        valor = linea[i]
        if valor == '0':
            contador_cero += 1
        if valor == '1':
            contador_uno += 1
    if contador_cero > contador_uno:
        gamma = gamma + '0'
        epsilon = epsilon + '1'
    else:
        gamma = gamma + '1'
        epsilon = epsilon + '0'
    contador_cero = 0
    contador_uno = 0

decimal_gamma = binario_a_decimal(gamma)
decimal_epsilon = binario_a_decimal(epsilon)
resultado = decimal_gamma * decimal_epsilon

print(resultado)

1082324


## Resultado Parte1: 1082324

## Parte2

--- La segunda parte ---
A continuación, debe verificar la clasificación de soporte vital, que se puede determinar multiplicando la clasificación del generador de oxígeno por la clasificación del depurador de CO2.

Tanto la clasificación del generador de oxígeno como la clasificación del depurador de CO2 son valores que se pueden encontrar en su informe de diagnóstico; encontrarlos es la parte difícil. Ambos valores se ubican mediante un proceso similar que implica filtrar valores hasta que solo queda uno. Antes de buscar cualquiera de los valores de calificación, comience con la lista completa de números binarios de su informe de diagnóstico y considere solo el primer bit de esos números. Luego:

Mantenga solo los números seleccionados por los criterios de bit para el tipo de valor de clasificación que está buscando. Descarte los números que no coincidan con los criterios de bits.
Si solo le queda un número, deténgase; este es el valor de clasificación que está buscando.
De lo contrario, repita el proceso, considerando el siguiente bit a la derecha.
Los criterios de bits dependen del tipo de valor de clasificación que desee encontrar:

Para encontrar la clasificación del generador de oxígeno, determine el valor más común (0 o 1) en la posición actual del bit y mantenga solo los números con ese bit en esa posición. Si 0 y 1 son igualmente comunes, mantenga los valores con un 1 en la posición que se está considerando.
Para encontrar la clasificación del depurador de CO2, determine el valor mínimo común (0 o 1) en la posición actual del bit y mantenga solo los números con ese bit en esa posición. Si 0 y 1 son igualmente comunes, mantenga los valores con un 0 en la posición que se está considerando.
Por ejemplo, para determinar el valor nominal del generador de oxígeno utilizando el mismo informe de diagnóstico de ejemplo anterior:

Comience con los 12 números y considere solo el primer bit de cada número. Hay más 1 bits (7) que 0 bits (5), así que mantenga solo los 7 números con un 1 en la primera posición: 11110, 10110, 10111, 10101, 11100, 10000 y 11001.
Luego, considere el segundo bit de los 7 números restantes: hay más 0 bits (4) que 1 bits (3), así que mantenga solo los 4 números con un 0 en la segunda posición: 10110, 10111, 10101 y 10000.
En la tercera posición, tres de los cuatro números tienen un 1, así que mantén esos tres: 10110, 10111 y 10101.
En la cuarta posición, dos de los tres números tienen un 1, así que mantén esos dos: 10110 y 10111.
En la quinta posición, hay un número igual de 0 bits y 1 bits (uno cada uno). Entonces, para encontrar la clasificación del generador de oxígeno, mantenga el número con un 1 en esa posición: 10111.
Como solo queda un número, deténgase; la clasificación del generador de oxígeno es 10111, o 23 en decimal.
Luego, para determinar el valor de clasificación del depurador de CO2 del mismo ejemplo anterior:

Comience de nuevo con los 12 números y considere solo el primer bit de cada número. Hay menos 0 bits (5) que 1 bits (7), así que mantenga solo los 5 números con un 0 en la primera posición: 00100, 01111, 00111, 00010 y 01010.
Luego, considere el segundo bit de los 5 números restantes: hay menos 1 bits (2) que 0 bits (3), así que mantenga solo los 2 números con un 1 en la segunda posición: 01111 y 01010.
En la tercera posición, hay un número igual de 0 bits y 1 bits (uno cada uno). Entonces, para encontrar la clasificación del depurador de CO2, mantenga el número con un 0 en esa posición: 01010.
Como solo queda un número, deténgase; la clasificación del depurador de CO2 es 01010, o 10 en decimal.
Finalmente, para encontrar la clasificación de soporte vital, multiplique la clasificación del generador de oxígeno (23) por la clasificación del depurador de CO2 (10) para obtener 230.

Utilice los números binarios en su informe de diagnóstico para calcular la clasificación del generador de oxígeno y la clasificación del depurador de CO2, luego multiplíquelos. ¿Cuál es la clasificación de soporte vital del submarino? (Asegúrese de representar su respuesta en decimal, no en binario).

In [51]:
def revisar_lista_oxigeno(lista, iterador):
    lista_oxigeno = []
    contador_cero = 0
    contador_uno = 0
    for item in lista:
        valor = item[iterador]
        if valor == '0':
            contador_cero += 1
        if valor == '1':
            contador_uno += 1
    if contador_cero > contador_uno:
        for item in lista:
            if item[iterador] == '0':
                lista_oxigeno.append(item)
    else:
        for item in lista:
            if item[iterador] == '1':
                lista_oxigeno.append(item)
    # lista_guarda = []
    return lista_oxigeno


def revisar_lista_depurador(lista, iterador):
    lista_depurador = []
    contador_cero = 0
    contador_uno = 0
    for item in lista:
        valor = item[iterador]
        if valor == '0':
            contador_cero += 1
        if valor == '1':
            contador_uno += 1
    if contador_cero < contador_uno:
        for item in lista:
            if item[iterador] == '0':
                lista_depurador.append(item)
    else:
        for item in lista:
            if item[iterador] == '1':
                lista_depurador.append(item)
    # lista_guarda = []
    return lista_depurador


contenido = open('input_puzzle3.txt', 'r')
mensaje = contenido.readlines()
resultado_1 =  []
resultado_2 = []
for linea in mensaje:
    resultado_1.append(linea.strip())
    resultado_2.append(linea.strip())

iterador_1 = 0
while iterador_1 < len(resultado_1[0]):
    resultado_1 = revisar_lista_oxigeno(resultado_1, iterador_1)
    iterador_1 += 1
print(resultado_1)
# print(resultado_2)
iterador_2 = 0
# print(iterador_2)
while iterador_2 < len(resultado_2[0]):
    resultado_2 = revisar_lista_depurador(resultado_2, iterador_2)
    iterador_2 += 1
print(resultado_2)




IndexError: list index out of range